src/modules/0.0.22/iasql_functions/iasql.ts

```diff
@@ -1037,3 +1039,633 @@ export async function continueUpgrade(
   });
   await install([...modsToInstall.values()], dbId, dbUser, false, true);
 }
+
+export async function commit(dbId: string, dryRun: boolean, context: Context) {
+  const t1 = Date.now();
+  logger.info(`Applying commit to ${dbId}`);
+  const versionString = await TypeormWrapper.getVersionString(dbId);
+  const importedModules = await getImportedModules(dbId, versionString);
+  let orm: TypeormWrapper | null = null;
+  try {
+    orm = await TypeormWrapper.createConn(dbId);
+    context.orm = orm;
+    const newStartCommit = await insertCommit(orm, dryRun ? 'preview_start' : 'start');
+    context.startCommit = newStartCommit;
+    const previousStartCommit = await getPreviousStartCommit(orm);
+    const changesToCommit: IasqlAuditLog[] = await orm.find(IasqlAuditLog, {
+      order: { ts: 'DESC' },
+      where: {
+        changeType: Not(In([AuditLogChangeType.START_COMMIT, AuditLogChangeType.END_COMMIT])),
+        ts: previousStartCommit
+          ? Between(previousStartCommit.ts, newStartCommit.ts)
+          : LessThan(newStartCommit.ts),
+        user: Not(config.db.user),
+      },
+    });
+    const tablesWithChanges = [...new Set(changesToCommit.map(c => c.tableName))];
+    const changesByEntity: { [key: string]: any[] } = getChangesByEntity(changesToCommit);
+
+    const installedModulesNames = (await orm.find(IasqlModule)).map((m: any) => m.name);
+    const installedModules: ModuleInterface[] = (Object.values(importedModules) as ModuleInterface[]).filter(
+      mod => installedModulesNames.includes(`${mod.name}@${mod.version}`),
+    );
+
+    const modulesWithChanges: ModuleInterface[] = getModulesWithChanges(
+      importedModules,
+      installedModules,
+      installedModulesNames,
+      tablesWithChanges,
+      versionString,
+    );
+    const t2 = Date.now();
+    logger.info(`Setup took ${t2 - t1}ms`);
+
+    const toCreate: Crupde = {};
+    const toUpdate: Crupde = {};
+    const toReplace: Crupde = {}; // Not actually used in sync mode, at least right now
+    const toDelete: Crupde = {};
+    if (modulesWithChanges.length) {
+      const rootToLeafOrder: ModuleInterface[] = sortModules(modulesWithChanges, installedModulesNames);
+      await commitSync(
+        dbId,
+        rootToLeafOrder,
+        context,
+        toCreate,
+        toUpdate,
+        toReplace,
+        toDelete,
+        dryRun,
+        changesByEntity,
+      );
+      return await commitApply(
+        dbId,
+        rootToLeafOrder,
+        context,
+        toCreate,
+        toUpdate,
+        toReplace,
+        toDelete,
+        dryRun,
+        changesByEntity,
+      );
+    } else {
+      /*
+       * If no changes found, what to do?
+       * If we sync everything and a user inserts something during the sync, could be detected and override the changes.
+       * We need to bring changes on AWS only always, ignore db only changes, but what to do with existing db records that might have changed? how to avoid overrides?
```

Comment:
Perhaps apply the filter on records to ignore based on the audit log to *both* cloud and db? If both changed simultaneously, we should still defer to the user's choice, right?

Reply:
oh, my bad to leave this misleading comment. I think I've solved this by [getting the latest changes](https://github.com/iasql/iasql-engine/pull/1453/files#diff-d1cec5ce18674d505cb6eddf5fe3c9421b5c2279e25df59363a3edd73060e80fR1282-R1283) every time we `db.read` during `commitSync` and then [filter the changes if necessary](https://github.com/iasql/iasql-engine/pull/1453/files#diff-d1cec5ce18674d505cb6eddf5fe3c9421b5c2279e25df59363a3edd73060e80fR1318-R1332)